capsules/extra/src/cycle_count.rs

```diff
@@ -0,0 +1,103 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Provides a cycle counter interface for userspace.
+//!
+//! Usage
+//! -----
+//!
+//! This capsule is intended for debug purposes. However, to ensure that use
+//! by multiple apps does not lead to innacurate results, basic virtualization
+//! is implemented: only the first app to start the cycle counter can start or
+//! stop or reset the counter. Other apps are restricted to reading the counter
+//! (which can be useful for debugging the time required by cross-process routines).
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::CycleCount as usize;
+
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::OptionalCell;
+use kernel::{hil, ErrorCode, ProcessId};
+
+#[derive(Default)]
+pub struct App;
+
+pub struct CycleCount<'a, P: hil::hw_debug::CycleCounter> {
+    counters: &'a P,
+    apps: Grant<App, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
+    controlling_app: OptionalCell<ProcessId>,
+}
+
+impl<'a, P: hil::hw_debug::CycleCounter> CycleCount<'a, P> {
+    pub fn new(
+        counters: &'a P,
+        grant: Grant<App, UpcallCount<0>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Self {
+        Self {
+            counters,
+            apps: grant,
+            controlling_app: OptionalCell::empty(),
+        }
+    }
+}
+
+impl<'a, P: hil::hw_debug::CycleCounter> SyscallDriver for CycleCount<'a, P> {
+    /// Control the CycleCount system.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Start the cycle counter.
+    /// - `2`: Get current cycle count.
+    /// - `3`: Reset and stop the cycle counter.
+    /// - `4`: Stop the cycle counter.
+    fn command(&self, command_num: usize, _data: usize, _: usize, app: ProcessId) -> CommandReturn {
```

Comment:
RISC-V supports exposing these directly to userspace to avoid the syscall overhead. It's controllable by the kernel and not always supported. It might be worth having some way to indicate to userspace if it can/can't access them?

That can be added later, but the HIL might want a `expose_to_userspace()` function

Reply:
I think that sounds like a good idea, but we have generally avoided adding methods to HILs until we have an implementation. I don't think there is anything stopping us from adding such a function once we have a RISC-V implementation